### ISSUE TITLE
Beta Splatting fine-tuning breaks down with sparse reconstructions - Any Insights?

### ISSUE THREAD

**Issue by rab-dd** [RongLiu-Leo/beta-splatting#15]
Hey! first of all, thank you so much for this great work.
SO I've been working with Gaussian Splatting for about 7 months now including Beta Splatting for the last month and ran into something weird that I can't figure out. Hoping you might have some insights.
So I had this dataset that worked great - 484 images of old European buildings with lots of texture detail. COLMAP gave me a nice dense point cloud (719 points per image), and after tuning the dead gaussiansn opacity threshold down from its default 0.005 to the range (0.001→0.00025) ,**which basically the only threshold I needed to tune**, 
`
dead_mask = (beta_model.get_opacity <= {value}).squeeze(-1)

beta_model.relocate_gs(dead_mask=dead_mask)
beta_model.add_new_gs(cap_max=args.cap_max)
`
Beta Splatting was producing beautiful results. This was especially important because Beta Splatting is literally the only method I've found that can handle tree branches properly along with high detailed surfaces and roads!

Now I'm working on a new dataset - 1,273 images of modern buildings with really detailed, thin trees everywhere. The problem is COLMAP is struggling hard with this one. The modern architecture is pretty uniform and those super thin tree branches just don't give enough features for COLMAP to work with. End result: only 212 points per image (basically 1/3 of what I had before).
Here's where it gets weird. I throw this sparse reconstruction at Convex Splatting and Hierarchical Splatting - they work fine! Stable training, decent results. But Beta Splatting? Complete disaster.
I've tried everything I can think of:
    • Dead Gaussian Opacity thresholds of 0.0008, 0.001, 0.002 and 0.005 - doesn't matter 
    • Different opacity learning rates (0.02, 0.05, 0.08) - zero difference 
    • Longer training times - actually makes it worse
The pattern is always the same: training starts okay (loss around 0.20), then just goes to hell. Loss climbs to 0.29+, the algorithm starts relocating millions of Gaussians per step (I've seen 5.4M relocations in one go!), and eventually it just crashes with OOM. It's like the algorithm gets stuck in some destructive loop where it keeps making things worse instead of better.
What's really puzzling me is why Beta Splatting specifically has this problem when the other methods handle the same sparse data just fine. Is there something different about how Beta Splatting handles densification or optimization that makes it need denser initialization?
This is pretty frustrating because Beta Splatting is my go-to for scenes with lots of vegetation, but now I can't use it on datasets where the preprocessing isn't perfect. Any ideas what might be causing this sensitivity to this dataset?
Thanks for any insights and many thanks for the work again!

**Comment by RongLiu-Leo** [comment]
Could you provide some visuals?
I'd love to help.
Try to use python train.py --share_url, which gives you a viewer link that could be visualized everywhere (Your phone, labtop, pad, etc.).
Possibly it would see the problem and I can address that.

**Comment by rab-dd** [comment]
Hey! first of all I'm so sorry for the late reply but I was trying to solve this issue.
Secondly, Thanks for offering to help with the Beta Splatting issue. Since I couldn't share the viewer link from Colab, I've rendered some scenes that show exactly what's happening.
Regarding my questions above about the role of the sparse point cloud in the method, I have already improved the sparse reconstruction from 212 to 396 points per image using sequential matching, but Beta Splatting still shows the exact same failure pattern. The weird thing is Convex and Hierarchical Splatting work fine on both sparse reconstructions as explained above.
I have analyzed the pattern of different training experiments and found really weird thing going  - there's a hardcoded 1.05 growth multiplier in the densification that just blindly adds 5% more Gaussians every step regardless of scene complexity. Plus the relocation system is basically just moving struggling Gaussians to random spots instead of actually removing them. That's why I keep seeing millions of relocations per step and the loss actually gets worse over time (If I have any mistakes in my understanding about the working mechanism, please correct them).

The attached scenes show the difference pretty clearly. With both sparse reconstructions, Beta Splatting produces these degraded results., the results shows the ground truth and the rendering after 3000, 5000 and 7500 iterations, also it is worth noting that the results were completely blind, just off-white color after the iteration 10000. Training logs show the same pattern every time - massive relocation thrashing and quality degradation.

I'm thinking of modifying the densification to use gradient-based decisions like the other methods do, and replace relocation with actual pruning. Does that sound like the right direction? Any other factors you think I should consider?

Really appreciate the help on this!

<img width="1846" height="302" alt="Image" src="https://github.com/user-attachments/assets/b9fa2d5d-2276-4fcd-8e85-10857ba2ac1f" />
<img width="1846" height="302" alt="Image" src="https://github.com/user-attachments/assets/d6a1616c-0fb2-4a16-936c-92cabec21c9f" />
<img width="1846" height="302" alt="Image" src="https://github.com/user-attachments/assets/6a205545-97a0-43ae-9e04-42df1bb9367c" />

**Comment by rab-dd** [comment]
<img width="4143" height="2738" alt="Image" src="https://github.com/user-attachments/assets/6bce2e44-2002-4b99-b530-df60a912896d" />
<img width="4143" height="2738" alt="Image" src="https://github.com/user-attachments/assets/c35c3bf5-b50f-4b58-a9ae-4eadb3075858" />
<img width="4143" height="2738" alt="Image" src="https://github.com/user-attachments/assets/060f5635-da04-408b-8ee0-9452d4eb844d" />

**Comment by RongLiu-Leo** [comment]
Actually this new densification strategy should be more efficient compared to the original one.
It relocates the dead primitives to the places that really need more primitive to construct. It is not a random place.

Some possible solutions:
1. If it has too many primitives relocated, may try to further improve the Dead Gaussian Opacity threshold to prevent relocating.
2. The new densification strategy will inject noise to primitive xyz location to encourage exploration. The noise is related to the scene scale. I guess this input scene has some floating point clouds that are far from the scene center, making the scene scale and noise unnecessarily large. Could you provide input.ply? And it is worthy to try specifying ```--noise_lr 0``` when training.

**Comment by rab-dd** [comment]
I'm training a small set of the data now, around 500 images and I will the mentioned noise learning rate, but can you please explain what do you mean by improving the dead gaussians opacity threshold? as I used values between (0.01 to 0.0001) and nothing worked.
for your second point, this could be right yeah, I have visualized the sparse point cloud and there are some floating points far from the scene, it was even clearer when using Hierarchcial gaussian splatting, one of the chunks include a scene that was far from the center.
but again, my question is how to improve the threshold of dead gaussians opacity? are there any factors that could tell me to increase or decrease the model?

**Comment by RongLiu-Leo** [comment]
So I think the major problem is due to the scene scale.
If this is the case, the dead primitive threshold is totally fine, try to reduce the noise_lr instead:
```
--noise_lr 1.0
```
or even lower
```
--noise_lr 0.1
```
```
--noise_lr 0.01
```
...
That should address your problem. 
Please reply this issue and I'd like to see if the problem would be addressed.

**Comment by rab306** [comment]
I'm ao sorry for my late response but I was working on another thing for the whole week. I will definitely try learning the whole scene again by reducing the noise learning and I will get back with the results. 
Thank you so much for your help

**Comment by rab306** [comment]
Also I want to mention that my experiment on the small set was successful but it's worth noting that it produced excellent results on the old dataset for the vegetation area and building roofs butit loses this advantage in this experiment despite the similarities between these objects in the datasets 
The difference I noticed is that, both scenes trained until reach 16,700,000 gaussians but the number of relocations in the old experiment (the excellent result) was smaller than the relocations in the current experiment. 
Is the number of relocations by the end of the training can be an indicator to the visual quality? In other words, can we set it that smaller relocations at end of training= better results?
